packages/sources/finalto/src/transport/forex.ts

```diff
@@ -0,0 +1,161 @@
+import { WebsocketReverseMappingTransport } from '@chainlink/external-adapter-framework/transports'
+import { BaseEndpointTypes } from '../endpoint/forex'
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+
+const logger = makeLogger('FinaltoWSTransport')
+
+export interface WSResponse {
+  MsgSeqNum: string
+  MsgType: string
+  Prices: {
+    Px: string
+    Type: 'B' | 'O'
+    Vol: string
+  }[]
+  SendTime: string
+  SubID: string
+  Symbol: string
+}
+
+export type WsTransportTypes = BaseEndpointTypes & {
+  Provider: {
+    WsMessage: WSResponse
+  }
+}
+
+let subscriptionId = 2 // DP errors if the subId starts with '1'
+const subscriptionIdMap: Record<string, number> = {}
+
+const getSubscriptionId = (symbol: string, deleteOnGet = false): string => {
+  if (subscriptionIdMap[symbol]) {
+    if (deleteOnGet) {
+      const value = subscriptionIdMap[symbol].toString()
+      delete subscriptionIdMap[symbol]
+      return value
+    }
+    return subscriptionIdMap[symbol].toString()
+  }
+  subscriptionIdMap[symbol] = subscriptionId
+  subscriptionId++
+  return subscriptionIdMap[symbol].toString()
+}
```

Comment:
Splitting this into two functions for getting and deleting might make it clearer what is going on - at first glance I wasn't sure what `deleteOnGet` meant

```suggestion
const getSubscriptionId = (symbol: string): string => {
  if (subscriptionIdMap[symbol]) {
    return subscriptionIdMap[symbol].toString()
  }
  subscriptionIdMap[symbol] = subscriptionId
  subscriptionId++
  return subscriptionIdMap[symbol].toString()
}

const deleteSubscriptionId = (symbol: string): string => {
      const value = subscriptionIdMap[symbol].toString()
      delete subscriptionIdMap[symbol]
      return value
}
```

Reply:
Actually now i don't think we even need `deleteSubscriptionId` function. Updated the code so that we delete the key in unsubscribeMessage directly after getting it.